### PR TITLE
ICL ICCMAX support

### PR DESCRIPTION
**Pull request by fredoh9**
Updated Nov 2,
First 3 commits are from PR 2297. Last a commit is for ICL ICCMAX support change.

Updated Nov 3,
First 3 commits are from PR 2297. Last four commits are for ICL ICCMAX support change.

For last commit, for explicit header inclusion. This is the right header required. This is valid style compliance fix per Documentation/process/submit-checklist.rst.
```
#include <linux/kernel.h>	// ARRAY_SIZE
#include <linux/kconfig.h>	// IS_ENABLED, CONFIG_
#include <linux/export.h>	// EXPORT_SYMBOL_
#include <linux/bits.h>		// GENMASK, BIT

#include <linux/module.h>	// MODULE_
```

Updated Nov 4,
Dropped symatrical changes for other dsp interface.
Re-ordered and combining commits. Now last two are ICL ICCMAX change.

### CODE REVIEW

File: sound/soc/sof/intel/icl.c
Comment (ranj063) [inline sound/soc/sof/intel/icl.c]: How am I the author for a file that you added?

File: sound/soc/sof/ops.h
Comment (ranj063) [inline sound/soc/sof/ops.h:112]: We don't need a new op for this right? We have core_power_up op and you can pass the core is of the last core

File: sound/soc/sof/intel/icl.c
Comment (ranj063) [inline sound/soc/sof/intel/icl.c]: What you really need is an ICL specific post_fw_run op. And call the coore_power_up op with the core mask for core 3 if hpro is enabled

File: sound/soc/sof/intel/icl.c
Comment (RanderWang) [inline sound/soc/sof/intel/icl.c]: Icelake

File: sound/soc/sof/intel/ext_manifest.h
Comment (RanderWang) [inline sound/soc/sof/intel/ext_manifest.h:22]: CAVS ?  Now I am working on SOF driver + CAVS FW, so I used it in my project, what is your CAVS ? thanks

Comment (kv2019i) [reply sound/soc/sof/intel/ext_manifest.h:22]: @RanderWang The existing naming has been that "cAVS" is name of the harwdware. For example: https://thesofproject.github.io/latest/platforms/intel-cavs/index.html

File: sound/soc/sof/intel/icl.c
Comment (kv2019i) [reply sound/soc/sof/intel/icl.c]: @ranj063 @fredoh9 Should the extra-core power-up be done already in snd_sof_dsp_run() .. i.e. is post_fw_run too late?

File: sound/soc/sof/sof-priv.h
Comment (kv2019i) [inline sound/soc/sof/sof-priv.h]: As commented elsewhere already, let's use existing ops to enable the additional core for iccmax.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @fredoh9 ! Could we run some stress tests whether this PR helps with https://github.com/thesofproject/sof/issues/3395 ? It would be really useful info and help to prioritize this PR.

**Comment by fredoh9** [comment]
> Thanks @fredoh9 ! Could we run some stress tests whether this PR helps with [thesofproject/sof#3395](https://github.com/thesofproject/sof/issues/3395) ? It would be really useful info and help to prioritize this PR.

I will run stress-test with this PR. I'm also curious about the result.

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [inline sound/soc/sof/intel/hda-loader.c]: two empty lines, will update in next time

File: sound/soc/sof/intel/ext_manifest.h
Comment (marc-hb) [reply sound/soc/sof/intel/ext_manifest.h:22]: I guess the name "cAVS" firmware made sense when it was the only one, before SOF existed.

Shameless plug: https://github.com/thesofproject/sof/issues/3491

**Comment by fredoh9** [comment]
I'm able to reproduce same error with the DEVICE TEST failure on ICL RVP. Looking at it.
```
jf-icl-rvp-hda-4 kernel: [    5.405310] sof-audio-pci 0000:00:1f.3: error: ipc timed out for 0x90030000 size 80
jf-icl-rvp-hda-4 kernel: [    5.405314] sof-audio-pci 0000:00:1f.3: info: preventing DSP entering D3 state to preserve context
jf-icl-rvp-hda-4 kernel: [    5.405335] sof-audio-pci 0000:00:1f.3: status: fw entered - code 00000005
jf-icl-rvp-hda-4 kernel: [    5.405512] sof-audio-pci 0000:00:1f.3: error: unexpected fault 0x00000000 trace 0x00004000
jf-icl-rvp-hda-4 kernel: [    5.405521] sof-audio-pci 0000:00:1f.3: error: hda irq intsts 0x00000000 intlctl 0xc0000000 rirb 00
jf-icl-rvp-hda-4 kernel: [    5.405522] sof-audio-pci 0000:00:1f.3: error: dsp irq ppsts 0x00000000 adspis 0x00000000
jf-icl-rvp-hda-4 kernel: [    5.405529] sof-audio-pci 0000:00:1f.3: error: host status 0x00000000 dsp status 0x00000000 mask 0x00000003
jf-icl-rvp-hda-4 kernel: [    5.405530] sof-audio-pci 0000:00:1f.3: error: can't set params for DMA for trace -110
jf-icl-rvp-hda-4 kernel: [    5.405556] sof-audio-pci 0000:00:1f.3: warning: failed to initialize trace -110
```

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: missing spaces around << and -?

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:501]: Can we please add a comment here saying that the recoemmended HW programming sequence for IC>L is to power up core 3 and keep it in stall if HPRO is enabled

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: not needed

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: we should be explicit in passing the core_mask to specify which core to stall here.

File: sound/soc/sof/intel/icl.c
Comment (ranj063) [inline sound/soc/sof/intel/icl.c]: Author

File: sound/soc/sof/intel/icl.c
Comment (lyakh) [inline sound/soc/sof/intel/icl.c:19]: you need more headers for `ID_ENABLED()`, `CONFIG_*`, `EXPORT_SYMBOL_NS()`, `GENMASK()`, `ARRAY_SIZE()` etc.

File: sound/soc/sof/intel/ext_manifest.h
Comment (kv2019i) [reply sound/soc/sof/intel/ext_manifest.h:22]: But the actual patch seems still correct. This is a config element type that is specific to firmware running on cAVS hardware (to separate from elements that are common to all hardware supported by SOF).

**Comment by kv2019i** [comment]
@fredoh9 wrote:
> I'm able to reproduce same error with the DEVICE TEST failure on ICL RVP. Looking at it.

Hmm, thanks a lot for running the tests. It's still possible there is some implementation issue in the flow (either this, or the FW part of ICCMAX flow), but at this point, much more likely scenario is that the boot failures are not linked to ICCMAX. In any case we need to use the correct flow, so this PR needs to proceed still.

File: sound/soc/sof/intel/ext_manifest.h
Comment (plbossart) [inline sound/soc/sof/intel/ext_manifest.h:1]: NAK on the file name.
You already have sound/sof/ext_manifest.h, changed in the previous patch.

This has got to be "intel_ext_manifest.h" or something clearly identifying Intel and not the generic SOF stuff

File: sound/soc/sof/intel/ext_manifest.h
Comment (plbossart) [inline sound/soc/sof/intel/ext_manifest.h:22]: what is the point of an EMPTY token that will be skipped anyways?

File: sound/soc/sof/intel/hda-loader.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c]: what does supported mean and who defines this?

File: sound/soc/sof/sof-priv.h
Comment (plbossart) [inline sound/soc/sof/sof-priv.h:104]: bisect issue, you need to change the caller in the same patch.

File: sound/soc/sof/intel/hda-loader.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c:513]: can you clarify why you need this routine when we don't seem to need it for TGL? Is this intentional?

File: sound/soc/sof/intel/apl.c
Comment (plbossart) [inline sound/soc/sof/intel/apl.c]: no, we keep the initial copyright.

File: sound/soc/sof/intel/byt.c
Comment (plbossart) [inline sound/soc/sof/intel/byt.c]: no

File: sound/soc/sof/intel/cnl.c
Comment (plbossart) [inline sound/soc/sof/intel/cnl.c]: no

**Review by plbossart — CHANGES_REQUESTED** [review]
Sorry, this isn't ready for merge yet.

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: lets use the top level core_power_up op and Bard's patch will take care of updating the mask automatically then

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: no need for this. lets just pass 1 << 3 directly for clarity

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: or BIT(3)

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:513]: @plbossart this is intentional. On ICL core 3 is managed by the host whereas on TGL it is handled by the firmware

Comment (plbossart) [reply sound/soc/sof/intel/hda-loader.c:513]: A comment wouldn't hurt then.

File: sound/soc/sof/intel/ext_manifest.h
Comment (lyakh) [reply sound/soc/sof/intel/ext_manifest.h:1]: @plbossart it is already in intel/ directory, can it really be confused for being SOF-generic? I find it a bit confusing in general: in most cases I prefer to eliminate redundancy in file names, so, sound/soc/sof/core.c is clear, that it's the SOF core, not ASoC or even ALSA core.c. But we also have sound/soc/sof/sof-audio.[ch], sof-priv.[ch], sof-{acpi,of,pci}-dev.c which seem both inconsistent and redundant, but ok, we already have them, and I personally don't see that as important enough to merit renames. OTOH ASoC has all their .c files named as sound/soc/soc-*.c which does look redundant to me, but at least it's consistent :-) In short, personally I find group/function.c preferable over group/group-function.c and it does seem to me the prevailing variant. But maybe this is a special case...

File: sound/soc/sof/intel/ext_manifest.h
Comment (lyakh) [reply sound/soc/sof/intel/ext_manifest.h:22]: @plbossart I think the explanation here was to check against 0-initialised data, as used in `switch` below, but I also don't find this neither useful nor common. There are *some* cases in the kernel where `enum`s skip 0 and begin with 1, but usually there's a very specific reason for that. Also the naming of `LAST_ELEM` which isn't in fact the last element but the one after it, usually named `NUMBER_OF_ELEMENTS` or similar. And using it as the size of an array in the ABI - all this I still don't like in this (and related) patch(es).

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c]: I didn't like this phase either but I copied from merged commit to get less review feedback.  I re-phased with what I feel right description

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:553]: separate patch please

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c:513]: I updated with more comments

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: this debug is not useful is it?

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: you can remove this variable and directly use GENMASK in the calls below maybe?

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c]: lets remove this. Bards PR will handle this when the enabled cores mask is updated in the cope_power_up op

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c]: I was wondering this code is kicked in with right core_mask. I can remove also.

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c]: ack!

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c:553]: ack! will update shortly again

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c]: I tried to use GENMASK(3,3) but there are multiple places to replace. I think define a variable and reuse it below is good idea.

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:565]: dont you need to check for core_mask and make sure they are in the host_managed_cores?

File: sound/soc/sof/intel/ext_manifest.h
Comment (marc-hb) [reply sound/soc/sof/intel/ext_manifest.h:1]: I don't like duplication in general either but note the number of "intel" occurrences in `#include "ext_manifest.h"` is zero, could be a bit confusing.

Now you also have to pray that the `Makefile`s know what they're doing which is quite rare :-) OK, hopefully not rare in the kernel...

Many editors also show only the basename in their menus etc.

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: sorry I meant BIT(3) not 1 << BIT(3)

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c:565]: ACK. Good point. Will include this.

File: sound/soc/sof/intel/hda-loader.c
Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c]: Yes, this was in between checking-in. I figured out it became high value, fixed to 1 << 3.
I will remove the variable and use BIT(3) only.

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:564]: do we need to check if core_mask is 0 and return if so?

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:553]: can this function be void?

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:569]: updates_bits only returns a true/false. I dont tink we need to return the value here. We should make it a void function or just return 0 after update_bits()

File: sound/soc/sof/intel/ext_manifest.h
Comment (fredoh9) [reply sound/soc/sof/intel/ext_manifest.h:1]: I like unique file name personally as I manage all the files in a project. But I already found multiple same file names in linux kernel directory. My understanding... if same file in `include directory`, there will be strong objection. linux kernel seems allow same file name in different `source directory` and doesn't like redundancy in file name. This file is in SOF `source directory`.
I don't have strong opinion. @plbossart what's your position? (after reading all comments after yours)

**Comment by ranj063** [comment]
@abonislawski Could you please take a look at the ICCMAX implementation for ICL here?

File: sound/soc/sof/intel/hda-loader.c
Comment (RanderWang) [inline sound/soc/sof/intel/hda-loader.c]: dev_info ? or return -EINVAL

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: not an error. just return 0 is enough

File: sound/soc/sof/intel/ext_manifest.h
Comment (fredoh9) [reply sound/soc/sof/intel/ext_manifest.h:22]: We need explicit EMPTY token to skip, I think, this is better than just comparing 0.
And also this is symmetrical change include/sound/sof/ext_manifest.h

https://github.com/thesofproject/linux/blob/topic/sof-dev/include/sound/sof/ext_manifest.h#L105
```
	SOF_EXT_MAN_CONFIG_EMPTY		= 0,
```

**Review by ranj063 — DISMISSED** [review]
LGTM. Thanks @fredoh9 for being patient.

**Review by abonislawski — APPROVED** [review]

File: sound/soc/sof/intel/hda-loader.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda-loader.c]: Note: I want to change this to dev_info(). If we have reached this point and we have an unknown header, it's a supported case of extending the ABI and not a cause for a warning. But not related to this patch.

File: sound/soc/sof/intel/Makefile
Comment (kv2019i) [inline sound/soc/sof/intel/Makefile:8]: Btw, @ranj063 and @fredoh9 -- while ICCMAX is somewhat known as a concept, in this case this is a very specific and non-obvious quirk of our DSP bootflow. I won't complain now as the same text is already merged for TGL in upstream, but the commit message is not very helpful for people outside SOF project. But yeah, I'm ok to go with this The bit about "recommended hw programming sequence" is key and it's there...

File: sound/soc/sof/intel/ext_manifest.h
Comment (kv2019i) [inline sound/soc/sof/intel/ext_manifest.h]: Did we conclude on the LAST_ELEM? I don't think you strictly need it in this patch as you are parsing based on IPC size and not based on size of sof_ext_man_cavs_config_data .

**Review by kv2019i — COMMENTED** [review]
Thanks @fredoh9 ! This indeed starts to look. I'll leave a bit more time to double-check before giving my approval. I left a few notes inline already now.

**Comment by fredoh9** [comment]
@ranj063 I lost your approval after addressing Kai's comment.
@plbossart @kv2019i do you have any comment or concern?

**Review by ranj063 — APPROVED** [review]

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @fredoh9 ! Code looks good now and seems to align as expected with merged FW.

I do not three warnings about lines over 100 columns in checkpatch. Please address these.

**Comment by fredoh9** [comment]
> I do not three warnings about lines over 100 columns in checkpatch. Please address these.

Sure, I will address that now!

**Review by kv2019i — APPROVED** [review]
Thanks @fredoh9 !

**Review by ranj063 — APPROVED** [review]

**Review by abonislawski — APPROVED** [review]

**Comment by kv2019i** [comment]
Ping @plbossart , you still have changes requested. Otherwise, I'm ready to press merge on this.

**Review by plbossart — APPROVED** [review]
@kv2019i I don't know what this fixes or improves, and the code looks good enough so I'll let you merge if you think it's fine.

**Pull request merged**
